okhttp/src/main/java/okhttp3/internal/connection/RealConnection.java

```diff
@@ -260,8 +260,16 @@ private void connectSocket(int connectTimeout, int readTimeout, Call call,
   private void establishProtocol(ConnectionSpecSelector connectionSpecSelector, Call call,
       EventListener eventListener) throws IOException {
     if (route.address().sslSocketFactory() == null) {
-      protocol = Protocol.HTTP_1_1;
       socket = rawSocket;
+
+      // this implementation takes the "prior knowledge" approach to the spec
+      // https://tools.ietf.org/html/rfc7540#section-3.4
+      if (route.address().protocols().contains(Protocol.H2C)) {
```

Comment:
I'd suggest we change 

https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/OkHttpClient.java#L854-L856

To allow H2C on it's own, but only on it's own, no fallback to HTTP/1.1.

Reply:
Done. This ended up cleaning up some of the logic in the establishment.